Support for external Usb devices

ysard opened the pull request:
I am currently working on a branch allowing the use of external USB modules (idea from #47, it could bring TVKILL to all Android devices).

They are produced by the company that publishes the ZaZa Remote application.
See [app](https://m.apkpure.com/fr/universal-tv-remote-zaza-remote/com.tiqiaa.remote).

- These are proprietary modules which were recently reversed here:

https://gitlab.com/XenRE/tiqiaa-usb-ir/
https://habr.com/ru/post/494800/

- The application is itself proprietary; I also broke my teeth during its reversing.

- The company publishes an SDK which allows its application and others like it, to query a large database of codes.
https://developer.tiqiaa.com/en/mainconsole.html
This SDK is licensed for a fee.

- Example of material (micro USB / USB-C):
https://fr.aliexpress.com/item/4000211545067.html


In short, these are fairly closed hardware and software. Encouraging the purchase of these modules may not be a good idea.
However, they cost around € 3 which is very honorable and is even more valuable if they can be hijacked.

---

My code and the lib integration interface correctly with TVKILL.
The Java / native code binding is done with the JNA project.
The integration of the native libusb library on Android is still imperfect and the example I had used JNA...

Personally, I would have preferred to use JNI directly or even convert the C++ code into pure Java for a lighter integration into the project. 
However, I am running out of time for these things...


In addition to the ethical aspects exposed above, I noticed that the sending of the codes is slowed down by approximately 6 seconds for 297 codes (for a duration of 64s instead of 58s).
The Java-JNA-C ++ additional cost therefore generates ~20ms per call.

---

The integration in TVKILL is finished but I'm still cleaning the lib code.

=> This PR is NOT ready and is based on my implementation of external json patterns (not currently merged).

I plan to integrate these projects (libusb and libtiqiaa) as submodules rather than hard in this repository. But this remains open.

What do you think of all this? :p

ysard committed:
Fix/update/add translations

ysard committed:
Implementation of custom JSON DB file; TODO: find a way to not restart app... #46

ysard committed:
Sync/add translations

ysard committed:
Bump to 1.4.0

ysard committed:
Merge branch 'dev' into custom_pattern_db

ysard committed:
Update translations: TV replaced by device

ysard committed:
Update translations: replace TV word by device

ysard committed:
Ability to auto-restart the app when a DB file is loaded; see c86e832

ysard committed:
Bump 1.5.0

ysard committed:
Remove unused strings

ysard committed:
Add doc: additional pattern option is confusing

ysard committed:
Fix confusing help about additional patterns: by default only the 1st signal of each brand is emitted

ysard committed:
Add few doc

ysard committed:
Uniformization of getProgressDialog: show() is called in it in all cases

ysard committed:
Remove unused import

ysard committed:
Follow #49; Fix JSON loading...
The filter on the mimetype application/json is in fact not supposed to work on Android which does not have
this entry in its DB. But It works against all odds on some platforms.
There is no way to my knowledge to detect this because MimeTypeMap does not contain this mimetype.
It is therefore necessary to let the user choose any file and then do the usual input verification operations to be sure to have JSON.

ysard committed:
Help some devices to open file picker in the folder containing the previous json file

ysard committed:
Fix potential error during restart with missing external db file

ysard committed:
Add support of USB Tiqiaa device
- USB auto detection
- Auto switch from internal IR blaster to external blaster
- Support for native lib based on libusb via JNA (native files included in following commit)
- Refactor Transmitter and Pattern classes to support these 2 modes of emission

ysard committed:
Fix typo

ysard committed:
Pattern convert fix: IR pattern should not be modified by this function and should only return converted pattern

ysard committed:
Add native wrapper files (src + config) for USB device

ysard commented:
Last minute note: I'm currently blocked by the lack of license of the code...
See: https://gitlab.com/XenRE/tiqiaa-usb-ir/-/issues/1

42SK commented:
This is very exiting!

Obviously, as you already mentioned, the main obstacle is the lack of a proper Licence for the library.

42SK commented:
This is very exiting!

Obviously, as you already mentioned, the main obstacle is currently the lack of a proper licence for the code you are using.
Maybe you could try sending the developer an e-mail kindly asking him to add a proper licence to his repository? (His e-mail address can be found in the copyright notice of the source files).
Alternatively, I guess one could always try to reimplement the driver, but that would probably be too much effort.

As for the advantages of JNI vs. JNA and the question if this should be its own submodule: I guess we should probably have a more thorough discussion about these things once we know that we can actually use the code.
Again, getting hold of the developer should be the top priority. It would be a shame if this would fail merely because of a trivial licensing issue.

ysard commented:
I sent a mail yesterday, we will see. It's a common situation...

Honestly the reimplementation from scratch would have been faster than the integration of the native code. My C is rusty and this is the first time I'm doing this in Java, despite the fact that the JNA seems to be an easier overlay to use.

The original application uses a native lib, but I think that it's more with the aim to mask a "trade secret" than for performance or feasibility reasons.

Alexamakans commented:
There was a response from the owner allowing use/modification of his code as long as a reference is kept to the original project.
Would be cool if you or anyone else is still interested in implementing this feature.

https://gitlab.com/XenRE/tiqiaa-usb-ir/-/issues/1#note_629481630

wasiliysoft commented:
I have just published an example of an Android application that does not use any third-party libraries, I need help to isolate the code into an independent library and test it. [https://github.com/wasiliysoft/tiqiaa-usb-demo-android](https://github.com/wasiliysoft/tiqiaa-usb-demo-android)

ysard committed:
Add tiqiaa usb files (YOLO import);
The author of tiqiaa lib agreed for the use of his work (sourced from gitlab repo).
